Path data guide (#33)
osg-htc/user-documentation: documentation/PATh/htc_workloads/using_data/data-transfer.md

---
path:
  path: htc_workloads/using_data/data-transfer.md
---

# Data Staging and Transfer to Jobs

Table of Contents: 
[TOC]

# Overview

> As a distributed system, jobs in the PATh Facility can run in different physical  locations, where the computers that are executing jobs don't have direct access to the files placed on the Access Point (e.g. in a home directory on ap1.facility.path-cc.io). In order to run on this kind of distributed system, jobs need to "bring along" the data, code, packages, and other files from the Access Point (where the job is submitted) to the PATh Facility execute points (where the job will run).  HTCondor's file transfer tools and plugins make this possible; input and output files are specified as part of the job submission and then moved to and from the execution location. 

Reviewer: ```suggestion
As a distributed system, jobs in the PATh Facility can run in different physical locations, where the computers that are executing jobs don't have direct access to the files placed on the Access Point (e.g. in a home directory on ap1.facility.path-cc.io). In order to run on this kind of distributed system, jobs need to "bring along" the data, code, packages, and other files from the Access Point (where the job is submitted) to the PATh Facility execute points (where the job will run).  HTCondor's file transfer tools and plugins make this possible; input and output files are specified as part of the job submission and then moved to and from the execution location. 
```

This guide describes where to place files on PATh Facility Access Points, and how to use these files within jobs. 

# Transferring Data To/From HTCondor Jobs

There are two spaces for placing files on the PATh Facility Access Point, and each has a corresponding transfer method for referencing files in the submit file. 

> <table>

Reviewer: I think we should simplify these tables. It is too easy to get into the "I have 1 50MB file, and one 500MB - what do I do?". My preference would be something like: <4GB total use CEDAR, >4GB total use stash

Author: I think those numbers have to be lower - 500MB instead of 4GB. 
Otherwise, I think that's okay?

Author: 1GB

<tr>
  <th>Location</th>
  <th>File Sizes</th>
  <th>Transfer Method</th>
  <th>Initial Quota</th>
</tr>
<tr>

>   <td rowspan="2"><code>/home/$USER</code></td>

Reviewer 2: Edit the table to have cutoffs of 1GB for input and output files

  <td>Input: less than 200Mb / file / job</td>
  <td rowspan="2">file paths in <code>transfer_input_files</code></td>
  <td rowspan="2">50GB</td>
</tr>
<tr>
  <td>Output: less than 1-2Gb / file / job</td>
</tr>
<tr>
  <td rowspan="2"><code>/data/$USER</code></td>
  <td>greater than 200Mb / file / job <br> OR shared files used by many jobs </td>
  <td rowspan="2"><code>stash:///</code> links in <code>transfer_input_files</code></td>
  <td rowspan="2">500GB / 250k items</td>
</tr>
<tr>
  <td> greater than 1-2Gb / file / job</td>
</tr>
</table>

Regardless of where data is placed, jobs should only be submitted with `condor_submit` from `/home`. 

## Transfer Smaller Job Input and Output Files to/from /home 

You should use your `/home` directory to stage job files where: 

> * individual input files per job are less than 200MB per file, and if there 

Reviewer 2: ```suggestion
* individual input files per job are less than 1GB per file, and if there 
```

> are multiple files, they total less than 500MB

Reviewer 2: ```suggestion
are multiple files, they total less than 1GB
```

> * output files per job are less than 1-2GB per file

Reviewer 2: ```suggestion
* output files per job are less than 1GB per file
```

Reviewer 2: How do we feel about this cutoff?

### Input Files from `/home`

To transfer input files from `/home`, list the files by name in the `transfer_input_files` submit file option. You can use either absolute or relative paths to your input files. Multiple files can be specified using a comma-separated list. 

Some examples: 

* Transferring multiple files from the submission directory
	```
	transfer_input_files = my_data.csv, my_software.tar.gz, my_script.py
	```

> * Transferring a file using an absolute path, useful if a file is not in the same directory tree as your submit file: 

Reviewer 2: ```suggestion
* Transferring a file using an absolute path is useful if a file is not in the same directory tree as your submit file: 
```

	```
	transfer_input_files = /home/username/path/to/my_software.tar.gz
	```

### Output Files to `/home`

By default, files created by your job will automatically be returned to your `/home` directory. If you would like a file to return to a diffrent subfolder within your `/home` directory, use HTCondor's `transfer_output_remaps` option. 

<!--For more information, see below.
Link to different guide??? --> 

## Transfer Larger Job Input and Output Files to/from `/path-facility/data`

You should use your `/path-facility/data` directory to stage job files where: 

> * individual input files per job are greater than 200MB per file

Reviewer 2: ```suggestion
* individual input files per job are greater than 1GB per file
```

* an input file (of any size) is used by many jobs

> * output files per job are greater than 1-2GB per file

Reviewer 2: ```suggestion
* output files per job are greater than 1GB per file
```

	```

# Moving Data to/from PATh Facility Access Points

In general, common Unix tools such as rsync, scp, PuTTY, WinSCP, gFTP, etc. can be used to upload data from your computer or another server to your PATh Facility Access Point or to download files. Files should be uploaded/created and staged in `/home` or `/path-facility` for preparation to use in jobs (as described above). 

# Check Your Quota and Available Space

## Check your `/home` quota

To check your home quota and usage, run: 

	$ quota -vs

##  Check your `/path-facility/data` quota

*coming soon!*

## Request Quota Increase

Contact us at support@path-cc.io if you think you need a quota increase. We have space for substantial workloads when communicated with in advance. 

# Data Policies

In general, users are responsible for managing data and for using appropriate mechanisms for delivering data to/from jobs. Each is controlled with a quota and should be treated as temporary storage for *active* job execution. The PATh Facility has no routine backup of data in these locations, and users should remove old data after jobs complete. 

> Data stored within `/home` and `/protected` is available only to your jobs, but highly sensitive data (e.g. HIPPA) should never be uploaded to OSG resources. 

Reviewer 2: ```suggestion
Data stored within `/home` and `/path-facility` is available only to your jobs, but highly sensitive data (e.g. HIPPA) should never be uploaded to OSG resources. 
```

PATh staff reserve the right to monitor and/or remove data without notice to the user if doing so is necessary for ensuring proper use or to quickly fix a performance or security issue. Additionally, users should not use PATh resources or services for long-term data storage (see above). 

> 

Reviewer: Maybe add something about none of this data being backed up?
